### DISCUSSION TITLE
Inquiries on creating UEFI capsule firmware

### DISCUSSION THREAD

**Discussion by ongeelim** [fwupd/fwupd#3415] [Q&A]
My team is currently working on adding fwupd component into a custom-made Linux OS as well as creating a UEFI capsule firmware file for our product. However, we would like to only distribute the capsule firmware in our own website.

I have a few questions that I would like to clarify:
1. From the list of metadata provided in https://lvfs.readthedocs.io/en/latest/metainfo.html, which metadata are considered essential for fwupd daemon?
2. It is stated in LVFS website that a GPG or PKCS#7 signature needs to be included in the cabinet file. Is there any significant impact if the signature file is not included and I just manually install it via "fwupdmgr install <.cab file>" command, especially in Secure Boot?
3. If GPG or PKCS#7 signature must be included, is there a way I could generate the signature internally instead of using LVFS server?

**Comment by hughsie** [comment]
> currently working on adding fwupd component into a custom-made Linux OS

Would that be a custom plugin too?

> However, we would like to only distribute the capsule firmware in our own website

Why?

> which metadata are considered essential

I think it's going to be much less work just adding all the metadata than working out if <project_license> is _actually_ required for instance.

> Is there any significant impact if the signature file is not included

You'll have to authenticate every time you deploy with the root password and you'll be warned the firmware is unsigned. Of course, if you have a custom Linux OS then you could just patch the polkit rule.

> is there a way I could generate the signature internally instead of using LVFS server

You can use `fwupdtool firmware-sign` but you'll also have to include the certificate you signed with as part of your OS image.

**Reply by ongeelim** [reply]
Thanks for the reply, Richard.
The reason why we are distributing the capsule firmware in our own website is that we're bundling that capsule firmware (in .cab file) with other software update components into a single package for distribution, and utilize that our own software to extract out the package and install that particular capsule firmware via fwupmgr command.

**Reply by superm1** [reply]
Does your product lock down the OS image and prevent replacing it with an off the shelf linux?  If not, then you should still consider to put the update in LVFS so that if anyone tries to do the update it works properly.  
Another bonus - if you have your image pull it from LVFS, all the signing infrastructure is already in place.
